backend/src/statistics/statistics.service.ts

```diff
@@ -0,0 +1,158 @@
+import { Injectable } from "@nestjs/common";
+import { differenceInSeconds, subYears } from "date-fns";
+
+import { PrismaService } from "src/core/prisma.service";
+
+import { UserStatisticsRes } from "~shared/types/api";
+import { Difficulty } from "~shared/types/base";
+
+@Injectable()
+export class StatisticsService {
+  constructor(private prisma: PrismaService) {}
+
+  async getByUserId(userId: number): Promise<UserStatisticsRes> {
+    const attemptSummaryPromise = this.getAttemptSummaryByUserId(userId);
+    const durationSummaryPromise = this.getDurationSummaryByUserId(userId);
+    const peerSummaryPromise = this.getPeerSummaryByUserId(userId);
+    const heatmapDataPromise = this.getHeatmapDataByUserId(userId);
+    const networkDataPromise = this.getNetworkDataByUserId(userId);
+
+    const attemptSummary = await attemptSummaryPromise;
+    const durationSummary = await durationSummaryPromise;
+    const peerSummary = await peerSummaryPromise;
+    const heatmapData = await heatmapDataPromise;
+    const networkData = await networkDataPromise;
+
+    return {
+      attemptSummary,
+      durationSummary,
+      peerSummary,
+      heatmapData,
+      networkData,
+    };
+  }
+
+  private async getAttemptSummaryByUserId(userId: number) {
+    const data = await this.prisma.roomSession.findMany({
+      where: { users: { some: { id: userId } }, endTime: { not: null } },
+      include: {
+        question: {
+          select: { difficulty: true },
+        },
+      },
+    });
+    const map = {
+      [Difficulty.EASY]: 0,
+      [Difficulty.MEDIUM]: 0,
+      [Difficulty.HARD]: 0,
+    };
+    for (const session of data) {
+      const { difficulty } = session.question;
+      map[difficulty] = map[difficulty] + 1;
+    }
+    return map;
+  }
+
+  private async getDurationSummaryByUserId(userId: number) {
+    const data = await this.prisma.roomSession.findMany({
+      where: {
+        users: { some: { id: userId } },
+        endTime: { not: null },
+        startTime: {
+          gte: subYears(new Date(), 1),
```

Comment:
Does this return one month's data? The heatmap appears to be using the exact same logic but yet is returning data for a year

Reply:
Fixed to return last 30 days intead.